Core/Object Arts/Dolphin/DolphinSure/BCryptLibrary.cls

```diff
@@ -0,0 +1,183 @@
+
+	NTSTATUS WINAPI BCryptHashData(
+		_Inout_	BCRYPT_HASH_HANDLE	hHash,
+		_In_		PUCHAR				pbInput,
+		_In_		ULONG				cbInput,
+		_In_		ULONG				dwFlags );"
+
+	<cdecl: dword BCryptHashData handle byte* dword dword>
+	^ self invalidCall.!
+
+openAlgorithmProvider: aHandle id: anId implementation: anImplementation flags: aFlags
+	"The BCryptOpenAlgorithmProvider function loads and initializes a BCrypt provider.
+
+	NTSTATUS WINAPI BCryptOpenAlgorithmProvider(
+		_Out_	BCRYPT_ALG_HANDLE*	phAlgorithm,
+		_In_		LPCWSTR				pszAlgId,
+		_In_		LPCWSTR				pszImplementation,
+		_In_		DWORD				dwFlags );"
+
+	<cdecl: dword BCryptOpenAlgorithmProvider handle* bstr bstr dword>
+	^ self invalidCall.!
+
+shaAlgorithmHandle
+	"Answer internal handle. to the SHA1 algorithm.
+	Create it, if it was not allocated yet.
+	This handle is never freed, because re-creating it is expensive."
+
+	| status |
+
+	( shaAlgorithmHandle isNil or: [ shaAlgorithmHandle value = 0 ] ) ifTrue: [
+		shaAlgorithmHandle := ExternalHandle new.
+		status := self openAlgorithmProvider: shaAlgorithmHandle id: BCRYPT_SHA1_ALGORITHM implementation: nil flags: 0.
+		( self statusFailed: status ) ifTrue: [ 
+			self error: 'BCrypt>>openAlgorithmProvider failed: ', ( status printStringRadix: 16 showRadix: false ) ] ].
+
+	^ shaAlgorithmHandle.!
+
+shaCreate
+	"Creates a SHA context. In BCrypt terms, the context is the hash handle."
+
+	| hashHandle status |
+
+	hashHandle := ExternalHandle new.
+	status := self createHash: self shaAlgorithmHandle hashHandle: hashHandle hashObject: nil hashObjectSize: 0 secret: 0 secretSize: 0 flags: 0.
+	( self statusFailed: status ) ifTrue: [ 
+		self error: 'BCrypt>>createHash: failed: ', ( status printStringRadix: 16 showRadix: false ) ].
+
+	^ hashHandle.!
+
+shaDestroy: aHashHandle
+	"Destroys the argument SHA handle (context)."
+
+	| status | 
+
+	status := self destroyHash: aHashHandle.
+	( self statusFailed: status ) ifTrue: [ 
+		self error: 'BCrypt>>destroyHash failed: ', ( status printStringRadix: 16 showRadix: false ) ].
+!
+
+shaGetHash: aHashHandle
+	"Answer the computed hash for the argument SHA handle.
+	 The answer will be a ByteArray of 20 bytes for SHA1"
+
+	| hash status |
+
+	hash := ByteArray new: 20.
+	status := self finishHash: aHashHandle output: hash outputSize: hash size flags: 0.
+	( self statusFailed: status ) ifTrue: [ 
+		self error: 'BCrypt>>finishHash failed: ', ( status printStringRadix: 16 showRadix: false ) ].
+
+	^ hash.!
+
+shaHashBuffer: aHashHandle buffer: buffer
+	"Hash the contents in of ByteArray buffer in the currently open hash context."
+
+	| status | 
+
+	status := self hashData: aHashHandle input: buffer inputSize: buffer size flags: 0.
+	( self statusFailed: status ) ifTrue: [ 
+		self error: 'BCrypt>>hashData failed: ', ( status printStringRadix: 16 showRadix: false ) ].
+!
+
+statusFailed: aStatus
+	"Answer true if the argument NTSTATUS indicates a fail value."
+
+	^ aStatus >= BCRYPT_STATUS_UNSUCCESSFUL.! !
```

Comment:
This isn't quite right. Any NTSTATUS with the high bit set is an error code. The cheapest way to detect this is to return an signed int and check for < 0.

Reply:
Implemented this as a compare with 16r80000000 (so no type conversions).